pkg/kubectl/cmd/portforward.go

```diff
@@ -113,27 +136,92 @@ func (f *defaultPortForwarder) ForwardPorts(method string, url *url.URL, opts Po
 	return fw.ForwardPorts()
 }
 
+// Translates service port to target port
+func translateServicePortToTargetPort(ports []string, mapping map[int32]int32) []string {
+	strmapping := make(map[string]string, len(mapping))
+	for k, v := range mapping {
+		strmapping[fmt.Sprintf("%d", k)] = fmt.Sprintf("%d", v)
+	}
+
+	var translated []string
+	for _, port := range ports {
+		// port is in the form of [LOCAL PORT]:REMOTE PORT
+		parts := strings.Split(port, ":")
+		switch len(parts) {
+		case 2:
+			if target, ok := strmapping[parts[1]]; ok {
+				translated = append(translated, fmt.Sprintf("%s:%s", parts[0], target))
+			} else {
+				translated = append(translated, port)
+			}
+		case 1:
+			if target, ok := strmapping[parts[0]]; ok {
+				translated = append(translated, fmt.Sprintf("%s:%s", parts[0], target))
+			} else {
+				translated = append(translated, port)
+			}
+		}
+	}
+	return translated
+}
+
 // Complete completes all the required options for port-forward cmd.
 func (o *PortForwardOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, args []string) error {
 	var err error
-	o.PodName = cmdutil.GetFlagString(cmd, "pod")
-	if len(o.PodName) == 0 && len(args) == 0 {
-		return cmdutil.UsageErrorf(cmd, "POD is required for port-forward")
+	if len(args) < 2 {
+		return cmdutil.UsageErrorf(cmd, "TYPE/NAME and list of ports are required for port-forward")
 	}
 
-	if len(o.PodName) != 0 {
-		printDeprecationWarning("port-forward POD", "-p POD")
-		o.Ports = args
-	} else {
-		o.PodName = args[0]
-		o.Ports = args[1:]
+	o.Namespace, _, err = f.DefaultNamespace()
+	if err != nil {
+		return err
 	}
 
-	o.Namespace, _, err = f.DefaultNamespace()
+	builder := f.NewBuilder().
+		Internal().
+		ContinueOnError().
+		NamespaceParam(o.Namespace).DefaultNamespace()
+
+	getPodTimeout, err := cmdutil.GetPodRunningTimeoutFlag(cmd)
+	if err != nil {
+		return cmdutil.UsageErrorf(cmd, err.Error())
+	}
+
+	resourceName := args[0]
+	builder.ResourceNames("pods", resourceName)
+
+	obj, err := builder.Do().Object()
+	if err != nil {
+		return err
+	}
+
+	forwardablePod, err := f.AttachablePodForObject(obj, getPodTimeout)
 	if err != nil {
 		return err
 	}
 
+	o.PodName = forwardablePod.Name
+
+	if cmdutil.GetFlagBool(cmd, "disable-target-port-translation") == false {
```

Comment:
I'd avoid adding the translation flag until we have a clear need for it

Reply:
Was thinking to have an escape hatch with this flag, thus default to false, in case there are use cases that I can't conceive.  I have no problem removing it and added it back later if needed.  I will do that also.

Reply:
port-forwarding to service is new function, so let's start small and iterate as needed... someone can always go look the pod up themselves if they want to target a specific port on the pod

Reply:
Removed.